### DISCUSSION TITLE
Weird performance issues, specifically at certain zoom levels and with certain combinations of edges. Seems like it can be fixed by using 3d transforms to force hardware acceleration

### DISCUSSION THREAD

**Discussion by Zombiefruit** [xyflow/xyflow#4617] [General]
Hey, sorry for the long winded title. I've been looking into performance issues we've been having with medium-sized (~50 nodes) and up reactflow graphs. At first I thought it was a problem with our custom nodes/edges, but upon closer inspection it seems like it actually might be a combination of how the browser handles transforming certain svgs at certain zoom levels (also, in certain browsers). Also, this may just be something specific to my machine (M1 Pro) - would love to see if other people experience the slowdown as well.

I'll try to keep it short. I've made a example sandbox [here](https://codesandbox.io/p/sandbox/zen-sid-pcgzvs?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0kx3zz100073b6iucifglqu%2522%252C%2522sizes%2522%253A%255B100%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0kx3zz100033b6idvc0af0o%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0kx3zz100043b6ikmuig3qt%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0kx3zz100063b6i6dahsokg%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0kx3zz100033b6idvc0af0o%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0kx3zz100023b6iyr6uxxgb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A48%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A48%252C%2522endColumn%2522%253A20%257D%255D%257D%255D%252C%2522id%2522%253A%2522cm0kx3zz100033b6idvc0af0o%2522%252C%2522activeTabId%2522%253A%2522cm0kx3zz100023b6iyr6uxxgb%2522%257D%252C%2522cm0kx3zz100063b6i6dahsokg%2522%253A%257B%2522id%2522%253A%2522cm0kx3zz100063b6i6dahsokg%2522%252C%2522activeTabId%2522%253A%2522cm0kxho4k000x3b6i8z4k003z%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0kx3zz100053b6iu5nkoqu2%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522cm0kx6cil001h3b6i63bponkn%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522cm0kxho4k000x3b6i8z4k003z%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm0kx3zz100043b6ikmuig3qt%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cm0kx3zz100043b6ikmuig3qt%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) for a demonstration of the issue ([sandbox preview without code](https://pcgzvs.csb.app/)).

Panning (for me) at the furthest zoom renders nicely, with an fps around 120, but zooming in (even just a tiny bit) results in an immediate fps drop to half of that. I believe it has something to do with all those layered edges. 

https://github.com/user-attachments/assets/7cee08ce-3a3c-4a5b-946d-fd8629e0c894

Pretty bizarre. What fixes this on my end is replacing the `translate` and `scale` transforms [here](https://github.com/xyflow/xyflow/blob/main/packages/react/src/container/Viewport/index.tsx#L6) with `translate3d` and `scale3d` to force hardware acceleration. After:

https://github.com/user-attachments/assets/1a2e7fb1-f78c-452a-83e8-e3c3004b7bbe

Note, I've only changed the `translate` transform here, not scale. Also, this example works fine for me on safari without the change, but lags even worse on Firefox. I'm wondering if it has something to do with how the different browsers handle svg overlap or decide to use hardware acceleration. I've dug around a bit but I couldn't come up with much. 

So I'm just curious, was the use of `translate` rather than `translate3d` a purposeful one? It seems to me like graphs like this are a good case for forcing hardware acceleration. Is this even the issue here? It's very strange that performance only drops for me at certain zoom levels. I'm no svg expert so this is really just my two cents. Looking at the code, it doesn't seem like there's an easy way to inject a different transform from an end user point of view, so I suppose I'll have to find a workaround. Would love to hear your thoughts though.

**Comment by moklick** [comment]
Hey @Zombiefruit ! Thanks for your input here. I had no luck with using `translate3d` for improving the performance for edge rendering. I know that certain styles like opacity have a huge impact when used for edges. Besides that, it seems that complex paths that overlap are always slow down a flow a lot. @peterkogo started to work on a prototype for a canvas renderer for edges. I could imagine that this could solve use cases like this. We will look into this again!

**Comment by VargasDevelopment** [comment]
Hey! SvelteFlow user here. I noticed the same issue and came across this thread.

In my case, just adding this to my css
```css
.svelte-flow__viewport {
  will-change: transform;
}
```
Fixed my performance issues. 

So I made this dynamic using SvelteFlow's built in [OnMove](https://svelteflow.dev/api-reference/types/on-move)

Here's the functions I used:
```ts
    let interactionTimer: NodeJS.Timeout;
    const interactionDelay = 150; // milliseconds 
    function addWillChange(event: { event: MouseEvent | null }) {
      const viewport:HTMLElement | null = document.querySelector('.svelte-flow__viewport'); 
      if(viewport){
        viewport.style.willChange = 'transform';
      }
      
      clearTimeout(interactionTimer);
    }

    function removeWillChange(event: { event: MouseEvent | null }) {
      const viewport:HTMLElement | null = document.querySelector('.svelte-flow__viewport'); 
      if(viewport){
        interactionTimer = setTimeout(() => {
          viewport.style.willChange = 'auto';
        }, interactionDelay);
      }
    }
```
And then I set these in my SvelteFlow component.

Now I get buttery smooth pan and zoom, and my nodes aren't pixelated when I zoom in because I remove the styling when I stop panning or zooming. Leaving this comment in hopes this will help someone else.
